core/src/jsMain/kotlin/Peripheral.kt

```diff
@@ -128,53 +133,105 @@ public class JsPeripheral internal constructor(
     private var connectJob: Deferred<Unit>? = null
 
     private fun connectAsync() = scope.async(start = LAZY) {
-        logger.info { message = "Connecting" }
-        _state.value = State.Connecting.Bluetooth
-
         try {
-            registerDisconnectedListener()
-
-            gatt.connect().await() // todo: Catch appropriate exception to emit State.Rejected.
-            _state.value = State.Connecting.Services
-
-            discoverServices()
-            onServicesDiscovered(ServicesDiscoveredPeripheral(this@JsPeripheral))
-            _state.value = State.Connecting.Observes
-            logger.verbose { message = "Configuring characteristic observations" }
-            observers.onConnected()
+            openConnection()
         } catch (t: Throwable) {
             logger.error(t) { message = "Failed to connect" }
-            disconnectGatt()
+            scope.async {
+                disconnectSync(cancelConnectJob = false)
+            }
             throw t
         }
-
-        logger.info { message = "Connected" }
-        _state.value = State.Connected
     }
 
-    private fun closeConnection() {
-        observationListeners.clear()
-        disconnectGatt()
-        unregisterDisconnectedListener()
-        _state.value = State.Disconnected()
+    private var disconnectJob: Deferred<Unit>? = null
+
+    private fun disconnectAsync(
+        cancelConnectJob: Boolean,
+    ) = scope.async {
+        if (cancelConnectJob) {
+            connectJob?.cancelAndJoin()
+        } else {
+            connectJob?.join()
+        }
+        closeConnection()
     }
 
     public override suspend fun connect() {
-        val job = connectJob ?: connectAsync().also { connectJob = it }
-        job.await()
+        connectSync()
     }
 
     public override suspend fun disconnect() {
-        job.cancelAndJoinChildren()
-        connectJob = null
-        disconnectGatt()
+        disconnectSync(cancelConnectJob = true)
+    }
+
+    private suspend fun connectSync() {
+        val job = connectJob ?: connectAsync().apply {
+            connectJob = this
+            invokeOnCompletion { connectJob = null }
+        }
+        job.await()
+    }
```

Comment:
This appears to change behavior in a way that isn't consistent with the [`connect` documentation](https://juullabs.github.io/kable/core/com.juul.kable/-peripheral/connect.html):

> If already connected, then returns immediately.

Before, we'd `await` the existing connect job, which after being connected, will always just return immediately because `await` on a completed `Deferred` immediately returns the result (i.e. would not attempt to reconnect when the connection previously completed successfully).

Whereas, this clears the `connectJob` upon successful connection, meaning subsequent calls will attempt to connect again, even though we're already connected.

Reply:
The problem with the previous implementation is that after a failed connection attempt, all future connection attempts would return an immediate failure on the `await` call, so it becomes impossible to retry the connection. This new implementation attempts to rectify that.

The `openConnection` method first checks if already connected and immediately returns. It is perhaps slightly inefficient to spin up the async job to do so, but it does satisfy the contract as documented.